homeassistant/components/abode/config_flow.py

```diff
@@ -26,48 +27,151 @@ def __init__(self):
             vol.Required(CONF_PASSWORD): str,
         }
 
+        self.mfa_data_schema = {
+            vol.Required(CONF_MFA): str,
+        }
+
+        self._username = None
+        self._password = None
+        self._polling = None
+        self._cache = None
+
+    async def _async_abode_login(self, step_id):
+        """Handle login with Abode."""
+        self._cache = self.hass.config.path(DEFAULT_CACHEDB)
+        errors = {}
+
+        try:
+            abode = Abode(
+                auto_login=False,
+                get_devices=False,
+                get_automations=False,
+                cache_path=self._cache,
+            )
+            await self.hass.async_add_executor_job(
+                abode.login, self._username, self._password
+            )
+
+        except (AbodeException, ConnectTimeout, HTTPError) as ex:
+            if ex.errcode == MFA_CODE_REQUIRED[0]:
+                return await self.async_step_mfa()
+
+            LOGGER.error("Unable to connect to Abode: %s", ex)
+
+            if ex.errcode == HTTP_BAD_REQUEST:
+                errors = {"base": "invalid_auth"}
+            else:
+                errors = {"base": "cannot_connect"}
+
+        if errors:
+            return self.async_show_form(
+                step_id=step_id, data_schema=vol.Schema(self.data_schema), errors=errors
+            )
+
+        return await self._async_create_entry(
+            {
+                CONF_USERNAME: self._username,
+                CONF_PASSWORD: self._password,
+                CONF_POLLING: self._polling,
+            }
+        )
+
+    async def _async_create_entry(self, user_input):
+        """Create the config entry."""
+        existing_entry = await self.async_set_unique_id(self._username)
+
+        if existing_entry:
+            self.hass.config_entries.async_update_entry(existing_entry, data=user_input)
+            # Force reload the Abode config entry otherwise devices will remain unavailable
+            self.hass.async_create_task(
+                self.hass.config_entries.async_reload(existing_entry.entry_id)
+            )
+
+            return self.async_abort(reason="reauth_successful")
+
+        return self.async_create_entry(title=self._username, data=user_input)
+
     async def async_step_user(self, user_input=None):
         """Handle a flow initialized by the user."""
         if self._async_current_entries():
             return self.async_abort(reason="single_instance_allowed")
```

Comment:
I noticed this method in the parent `ConfigFlow` class that some integrations use:

`_abort_if_unique_id_configured`

Is this the preferred method? As far as I can tell, it's essentially doing the same thing but with some more options (e.g. ability to update the existing flow).

Reply:
I guess one different might be with Abode, we only want to allow users to add one Abode account where as other integrations might allow users to add multiple accounts from an integration. I haven't been able to determine if it's possible with Abode yet but I just don't see it being something we need.

Reply:
That abort helper will have a standard abort message. We want to use another message when reauthenticating.

Reply:
This is just to check if the Abode integration is already set up. If it is, immediately abort the flow so the user can't set it up. After looking at it a little closer, it looks like `_abort_if_unique_id_configured` is only checking to see if that specific config flow already exists. Our case is a little different with Abode as we want to check to see if the integration has already been added, which I don't see a built in helper method for so I think this is fine.

Reply:
`_abort_if_unique_id_configured` will check for config entries with same unique_id as the current flow and abort the flow if there's an existing entry that matches.